src/panel/applets/status/PowerIndicator.vala

```diff
@@ -146,11 +146,130 @@ public class BatteryIcon : Gtk.Box {
+	public abstract string active_profile { owned get; set; }
+}
+
+public class PowerProfilesSelector : Gtk.Box {
+	// Power profile toggles
+	private Gtk.RadioButton radio_power_save;
+	private Gtk.RadioButton radio_power_balanced;
+	private Gtk.RadioButton radio_power_performance;
+
+	// these are used to temporarily block signals and prevent an infinite loop
+	private ulong radio_power_save_id;
+	private ulong radio_power_balanced_id;
+	private ulong radio_power_performance_id;
+
+	public PowerProfilesSelector(PowerProfilesDBus profiles_proxy) {
+		orientation = Gtk.Orientation.VERTICAL;
+		
+		var profiles = new GenericSet<string>(str_hash, str_equal);
+
+		foreach (HashTable<string, Variant> profile in profiles_proxy.profiles) {
+			var profile_value = profile.get("Profile");
+
+			if (!profile_value.is_of_type(VariantType.STRING)) {
+				warning("unexpected type for power profile: is not of VariantType.STRING");
+				continue;
+			}
+			
+			profiles.add(profile_value.get_string());
+		}
+		
+		if (profiles.length < 2) {
+			// need at least two options for it to be meaningful
+			warning("Fewer than two power profile options");
+			return;
+		}
+
+		// initialize state
+		on_active_profile_changed(profiles_proxy.active_profile);
+
+		((DBusProxy) profiles_proxy).g_properties_changed.connect(() => {
+			on_active_profile_changed(profiles_proxy.active_profile);
+		});
+
+		var sep = new Gtk.Separator(Gtk.Orientation.HORIZONTAL);
+		pack_start(sep, false, false, 1);
+
+		var header = new Gtk.Label("");
+		header.set_markup("<b>%s</b>".printf(_("Power Mode")));
+		header.set_halign(Gtk.Align.CENTER);
+		pack_start(header);
+
+		var power_profiles_radio_box = new Gtk.Box(Gtk.Orientation.HORIZONTAL, 1);
+
+		if (profiles.contains("power-saver")) {
+			radio_power_save = new Gtk.RadioButton.with_label_from_widget(null, _("Power Saver"));
+			radio_power_save_id = radio_power_save.toggled.connect(() => {
+				if (radio_power_save.get_active()) {
+					profiles_proxy.active_profile = "power-saver";
+				}
+			});
+			power_profiles_radio_box.pack_start(radio_power_save, false, false, 1);
+		}
+
+		if (profiles.contains("balanced")) {
+			radio_power_balanced = new Gtk.RadioButton.with_label_from_widget(radio_power_save, _("Balanced"));
+			radio_power_balanced_id = radio_power_balanced.toggled.connect(() => {
+				if (radio_power_balanced.get_active()) {
+					profiles_proxy.active_profile = "balanced";
+				}
+			});
+			power_profiles_radio_box.pack_start(radio_power_balanced, false, false, 1);
+		}
+
+		if (profiles.contains("performance")) {
+			radio_power_performance = new Gtk.RadioButton.with_label_from_widget(radio_power_save, _("Performance"));
+			radio_power_performance_id = radio_power_performance.toggled.connect(() => {
+				if (radio_power_performance.get_active()) {
+					profiles_proxy.active_profile = "performance";
+				}
+			});
+			power_profiles_radio_box.pack_start(radio_power_performance, false, false, 1);
+		}
```

Comment:
I think a fair bit of this code is duplicated and we can instead create class that takes the power proxy, type, etc. and creates a RadioButton that when toggled, fires off the type to the provided proxy.